gcc/d/dfrontend/expression.c

```diff
@@ -4082,6 +4082,12 @@ Expression *ArrayLiteralExp::semantic(Scope *sc)
         return new ErrorExp();
     }
 
+    if (global.params.noTypeinfo && elements->dim != 0)
+    {
+        error (global.params.noTypeinfo, "Can't use array literal");
```

Comment:
The problem here is that `getTypeInfo` might generate the TypeInfo for a type and this requires the TypeInfo declarations from `object.d` (it seems these are not needed for ctfe though). So we have to avoid calling `getTypeInfo`. We could also error in `getTypeInfo` but we can't provide good error messages in that case. I hoped I could return some dummy/error value and handle the error reporting in the visitor, but that doesn't work.

However, the visitor can only provide a small part of nortti anyway: There are still changes in `declaration.c`, `func.c`, `mtype.c` and `clone.c`.

>TypeInfo generation should probably not be done during/called from semantic anyway IMO.

We could probably need a post-semantic phase ;-) I think it'd also be nice to have the druntime-call-generation shared between all compilers but that can't be done in semantic, AFAIK.

It's indeed not possible to know whether `ArrayLiteralExp` is used for a static array in `ArrayLiteralExp::semantic`. The `type` field is initially always a dynamic array and fixed up after `ArrayLiteralExp::semantic`. For now I've updated the old code and moved the array literal error to the glue layer. This is possible as in this special case the TypeInfo is first accessed/generated by the glue layer.